File: 0106_med_hot_construct-binary-tree-from-inorder-and-postorder-traversal/solution.py
# Definition for a binary tree node.
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

        
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        if len(inorder) == len(postorder) == 0:
            return None

        if len(inorder) == len(postorder) == 1:
            return TreeNode(inorder[0])
        
        node = TreeNode(postorder[-1])

        root_index = inorder.index(node.val)

        node.left = self.buildTree(inorder[:root_index], postorder[:root_index])
        node.right = self.buildTree(inorder[root_index+1:], postorder[root_index:-1])

        return node

File: 0106_med_hot_construct-binary-tree-from-inorder-and-postorder-traversal/solution.java
/**
 * Definition for a binary tree node.
 */

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}


class Solution {
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        if(inorder.length == 0 && postorder.length == 0) {
            return null;
        }
        if(inorder.length == 1 && postorder.length == 1) {
            return new TreeNode(inorder[0]);
        }

        int size = inorder.length;

        TreeNode node = new TreeNode(postorder[size-1]);

        // int leftCount = Arrays.index

        return node;
    }
}